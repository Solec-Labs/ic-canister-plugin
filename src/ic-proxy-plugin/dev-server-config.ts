import type { WebpackOptionsNormalized } from 'webpack';
import type { DfxNetworkConfig, DfxNetwork, DfxConfig } from '../common';
import {
  mergeProxyOptions,
  WebpackDevServerProxyOptions,
} from './dev-server-proxy-config';
import {
  mergeStaticOptions,
  WebpackDevServerStaticOptions,
} from './dev-server-static-config';

export interface WebpackDevServerOptions {
  [index: string]: unknown;
  static?: WebpackDevServerStaticOptions;
  proxy?: WebpackDevServerProxyOptions;
  output?: WebpackOptionsNormalized['output'];
}

export function mergeDevServerConfig(
  webpackOptions: WebpackOptionsNormalized | undefined,
  dfxConfig: DfxConfig,
  dfxNetwork: DfxNetwork,
  dfxNetworkConfig: DfxNetworkConfig,
): WebpackDevServerOptions {
  const webpackDevServerOptions: WebpackDevServerOptions =
    webpackOptions?.devServer ?? {};

  const devServerStaticOptions = mergeStaticOptions(
    dfxConfig,
    webpackOptions?.output ?? {},
    webpackDevServerOptions.static,
  );
  const devServerProxyOptions = mergeProxyOptions(
    webpackDevServerOptions.proxy ?? {},
    dfxNetwork,
    dfxNetworkConfig,
  );

  const devServerOptions = {
    ...webpackDevServerOptions,
    static: devServerStaticOptions,
    proxy: devServerProxyOptions,
  };

  return devServerOptions;
}
