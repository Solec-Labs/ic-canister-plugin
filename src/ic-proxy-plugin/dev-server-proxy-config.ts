import { DfxNetworkConfig, DfxNetwork, getDfxNetworkAddress } from '../common';

export type WebpackDevServerProxyOptions = Record<string, unknown>;

export function mergeProxyOptions(
  originalProxyOptions: WebpackDevServerProxyOptions,
  dfxNetwork: DfxNetwork,
  dfxNetworkConfig: DfxNetworkConfig,
): WebpackDevServerProxyOptions {
  const networkBind = getDfxNetworkAddress(dfxNetwork, dfxNetworkConfig);

  return {
    '/api': {
      target: `http://${networkBind}`,
      changeOrigin: true,
      pathRewrite: {
        '^/api': '/api',
      },
    },
    ...originalProxyOptions,
  };
}
