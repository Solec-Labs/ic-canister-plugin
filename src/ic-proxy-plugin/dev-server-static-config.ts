import path from 'path';
import type { WebpackOptionsNormalized } from 'webpack';
import {
  DfxAssetCanisterConfig,
  DfxConfig,
  getDfxAssetCanisterConfig,
} from '../common';

export interface WebpackDevServerStaticFolder {
  directory: string;
}

export type WebpackDevServerStaticOptions =
  | boolean
  | string
  | string[]
  | WebpackDevServerStaticFolder
  | WebpackDevServerStaticFolder[]
  | (WebpackDevServerStaticFolder | string)[];

function getAssetCanisterStaticSources(
  outputOptions: WebpackOptionsNormalized['output'],
  assetCanisterConfig: DfxAssetCanisterConfig,
): WebpackDevServerStaticFolder[] {
  const outputPath = outputOptions.path ?? null;
  if (!outputPath) {
    return [];
  }

  const normalizedOutputPath = path.resolve(outputPath);
  const normalizedSources = assetCanisterConfig.source.map(source =>
    path.resolve(source),
  );

  return normalizedSources
    .filter(source => !source.startsWith(normalizedOutputPath))
    .map(source => ({
      directory: source,
    }));
}

export function mergeStaticOptions(
  dfxConfig: DfxConfig,
  outputOptions: WebpackOptionsNormalized['output'],
  originalStaticOptions: WebpackDevServerStaticOptions | undefined,
): WebpackDevServerStaticOptions {
  const assetCanisterConfig = getDfxAssetCanisterConfig(dfxConfig);

  const staticSources = getAssetCanisterStaticSources(
    outputOptions,
    assetCanisterConfig,
  );

  if (originalStaticOptions === false) {
    return false;
  }

  if (Array.isArray(originalStaticOptions)) {
    return [...originalStaticOptions, ...staticSources];
  }

  if (
    typeof originalStaticOptions === 'string' ||
    typeof originalStaticOptions === 'object'
  ) {
    return [originalStaticOptions, ...staticSources];
  }

  return staticSources;
}
