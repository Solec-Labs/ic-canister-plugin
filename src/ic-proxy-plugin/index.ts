export * from './ic-proxy-plugin';
