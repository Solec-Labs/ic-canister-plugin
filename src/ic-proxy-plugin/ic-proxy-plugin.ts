import type { Compiler } from 'webpack';
import { getDfxConfig, getDfxNetwork } from '../common';
import { mergeDevServerConfig } from './dev-server-config';

export class IcProxyPlugin {
  public apply(compiler: Compiler): void {
    const logger = compiler.getInfrastructureLogger('IcProxyPlugin');

    const dfxConfig = getDfxConfig();
    const dfxNetwork = getDfxNetwork(logger);
    const dfxNetworkConfig = dfxConfig.networks?.[dfxNetwork] ?? null;

    if (dfxNetworkConfig === null) {
      logger.log(
        `No network config found for ${dfxNetwork}, skipping webpack proxy setup`,
      );

      return;
    }

    compiler.hooks.afterEnvironment.tap('IcProxyPlugin', () => {
      compiler.options.devServer = mergeDevServerConfig(
        compiler.options,
        dfxConfig,
        dfxNetwork,
        dfxNetworkConfig,
      );
    });
  }
}
