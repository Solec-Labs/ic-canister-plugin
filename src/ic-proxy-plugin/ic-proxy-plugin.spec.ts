import { createMock } from 'ts-auto-mock';
import type { Compiler } from 'webpack';
import fs from 'fs';
import path from 'path';
import { IcProxyPlugin } from './ic-proxy-plugin';
import type { DfxConfig } from '../common';

jest.mock('fs');
jest.mock('path');

describe('IcProxyPlugin', () => {
  let plugin: IcProxyPlugin;

  let compilerMock: Compiler;
  let tapMock: jest.Mock;
  let readFileSpy: jest.SpyInstance;

  beforeEach(() => {
    readFileSpy = jest.spyOn(fs, 'readFileSync');
    jest.spyOn(path, 'resolve').mockImplementation(path => path);

    tapMock = jest.fn().mockImplementation((_, callback: () => void) => {
      callback();
    });
    compilerMock = createMock<Compiler>({
      options: {
        output: {
          path: path.resolve('./dist/hello_ic_frontend'),
        },
      },
      hooks: {
        afterEnvironment: {
          tap: tapMock,
        },
      },
    });
    plugin = new IcProxyPlugin();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should merge dev server config options', () => {
    const dfxConfig: DfxConfig = {
      canisters: {
        frontend: {
          type: 'assets',
          source: [
            'src/hello_ic_frontend/assets',
            'src/hello_ic_frontend/css',
            'dist/hello_ic_frontend/',
          ],
        },
        backend: {
          type: 'motoko',
        },
      },
      networks: { local: { bind: 'https://fake-network.com' } },
    };
    readFileSpy.mockReturnValue(Buffer.from(JSON.stringify(dfxConfig)));

    plugin.apply(compilerMock);

    expect(compilerMock.options.devServer).toMatchSnapshot();
  });

  it('should not do anything if there is no network config', () => {
    const dfxConfig: DfxConfig = {
      canisters: {
        frontend: {
          type: 'assets',
          source: [
            'src/hello_ic_frontend/assets',
            'src/hello_ic_frontend/css',
            'dist/hello_ic_frontend/',
          ],
        },
        backend: {
          type: 'motoko',
        },
      },
      networks: {},
    };
    readFileSpy.mockReturnValue(Buffer.from(JSON.stringify(dfxConfig)));

    plugin.apply(compilerMock);

    expect(compilerMock.options.devServer).toMatchSnapshot();
  });

  it('should not add static assets if there is no output path', () => {
    compilerMock = createMock<Compiler>({
      hooks: {
        afterEnvironment: {
          tap: tapMock,
        },
      },
    });

    const dfxConfig: DfxConfig = {
      canisters: {
        frontend: {
          type: 'assets',
          source: [
            'src/hello_ic_frontend/assets',
            'src/hello_ic_frontend/css',
            'dist/hello_ic_frontend/',
          ],
        },
        backend: {
          type: 'motoko',
        },
      },
      networks: { local: { bind: 'https://fake-network.com' } },
    };
    readFileSpy.mockReturnValue(Buffer.from(JSON.stringify(dfxConfig)));

    plugin.apply(compilerMock);

    expect(compilerMock.options.devServer).toMatchSnapshot();
  });

  it('should not add static assets if static assets is disabled', () => {
    compilerMock = createMock<Compiler>({
      options: {
        output: {
          path: path.resolve('./dist/hello_ic_frontend'),
        },
        devServer: {
          static: false,
        },
      },
      hooks: {
        afterEnvironment: {
          tap: tapMock,
        },
      },
    });

    const dfxConfig: DfxConfig = {
      canisters: {
        frontend: {
          type: 'assets',
          source: [
            'src/hello_ic_frontend/assets',
            'src/hello_ic_frontend/css',
            'dist/hello_ic_frontend/',
          ],
        },
        backend: {
          type: 'motoko',
        },
      },
      networks: { local: { bind: 'https://fake-network.com' } },
    };
    readFileSpy.mockReturnValue(Buffer.from(JSON.stringify(dfxConfig)));

    plugin.apply(compilerMock);

    expect(compilerMock.options.devServer).toMatchSnapshot();
  });

  it('should merge with existing static assets array', () => {
    compilerMock = createMock<Compiler>({
      options: {
        output: {
          path: path.resolve('./dist/hello_ic_frontend'),
        },
        devServer: {
          static: ['other-src/assets'],
        },
      },
      hooks: {
        afterEnvironment: {
          tap: tapMock,
        },
      },
    });

    const dfxConfig: DfxConfig = {
      canisters: {
        frontend: {
          type: 'assets',
          source: [
            'src/hello_ic_frontend/assets',
            'src/hello_ic_frontend/css',
            'dist/hello_ic_frontend/',
          ],
        },
        backend: {
          type: 'motoko',
        },
      },
      networks: { local: { bind: 'https://fake-network.com' } },
    };
    readFileSpy.mockReturnValue(Buffer.from(JSON.stringify(dfxConfig)));

    plugin.apply(compilerMock);

    expect(compilerMock.options.devServer).toMatchSnapshot();
  });

  it('should merge with existing static assets string', () => {
    compilerMock = createMock<Compiler>({
      options: {
        output: {
          path: path.resolve('./dist/hello_ic_frontend'),
        },
        devServer: {
          static: 'other-src/assets',
        },
      },
      hooks: {
        afterEnvironment: {
          tap: tapMock,
        },
      },
    });

    const dfxConfig: DfxConfig = {
      canisters: {
        frontend: {
          type: 'assets',
          source: [
            'src/hello_ic_frontend/assets',
            'src/hello_ic_frontend/css',
            'dist/hello_ic_frontend/',
          ],
        },
        backend: {
          type: 'motoko',
        },
      },
      networks: { local: { bind: 'https://fake-network.com' } },
    };
    readFileSpy.mockReturnValue(Buffer.from(JSON.stringify(dfxConfig)));

    plugin.apply(compilerMock);

    expect(compilerMock.options.devServer).toMatchSnapshot();
  });

  it('should merge with existing static assets object', () => {
    compilerMock = createMock<Compiler>({
      options: {
        output: {
          path: path.resolve('./dist/hello_ic_frontend'),
        },
        devServer: {
          static: {
            directory: 'other-src/assets',
          },
        },
      },
      hooks: {
        afterEnvironment: {
          tap: tapMock,
        },
      },
    });

    const dfxConfig: DfxConfig = {
      canisters: {
        frontend: {
          type: 'assets',
          source: [
            'src/hello_ic_frontend/assets',
            'src/hello_ic_frontend/css',
            'dist/hello_ic_frontend/',
          ],
        },
        backend: {
          type: 'motoko',
        },
      },
      networks: { local: { bind: 'https://fake-network.com' } },
    };
    readFileSpy.mockReturnValue(Buffer.from(JSON.stringify(dfxConfig)));

    plugin.apply(compilerMock);

    expect(compilerMock.options.devServer).toMatchSnapshot();
  });
});
