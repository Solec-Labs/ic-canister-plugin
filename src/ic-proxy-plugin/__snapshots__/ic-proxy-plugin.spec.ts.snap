// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IcProxyPlugin should merge dev server config options 1`] = `
{
  "proxy": {
    "/api": {
      "changeOrigin": true,
      "pathRewrite": {
        "^/api": "/api",
      },
      "target": "http://https://fake-network.com",
    },
  },
  "static": [
    {
      "directory": "src/hello_ic_frontend/assets",
    },
    {
      "directory": "src/hello_ic_frontend/css",
    },
    {
      "directory": "dist/hello_ic_frontend/",
    },
  ],
}
`;

exports[`IcProxyPlugin should merge with existing static assets array 1`] = `
{
  "proxy": {
    "/api": {
      "changeOrigin": true,
      "pathRewrite": {
        "^/api": "/api",
      },
      "target": "http://https://fake-network.com",
    },
  },
  "static": [
    "other-src/assets",
    {
      "directory": "src/hello_ic_frontend/assets",
    },
    {
      "directory": "src/hello_ic_frontend/css",
    },
    {
      "directory": "dist/hello_ic_frontend/",
    },
  ],
}
`;

exports[`IcProxyPlugin should merge with existing static assets object 1`] = `
{
  "proxy": {
    "/api": {
      "changeOrigin": true,
      "pathRewrite": {
        "^/api": "/api",
      },
      "target": "http://https://fake-network.com",
    },
  },
  "static": [
    {
      "directory": "other-src/assets",
    },
    {
      "directory": "src/hello_ic_frontend/assets",
    },
    {
      "directory": "src/hello_ic_frontend/css",
    },
    {
      "directory": "dist/hello_ic_frontend/",
    },
  ],
}
`;

exports[`IcProxyPlugin should merge with existing static assets string 1`] = `
{
  "proxy": {
    "/api": {
      "changeOrigin": true,
      "pathRewrite": {
        "^/api": "/api",
      },
      "target": "http://https://fake-network.com",
    },
  },
  "static": [
    "other-src/assets",
    {
      "directory": "src/hello_ic_frontend/assets",
    },
    {
      "directory": "src/hello_ic_frontend/css",
    },
    {
      "directory": "dist/hello_ic_frontend/",
    },
  ],
}
`;

exports[`IcProxyPlugin should not add static assets if static assets is disabled 1`] = `
{
  "proxy": {
    "/api": {
      "changeOrigin": true,
      "pathRewrite": {
        "^/api": "/api",
      },
      "target": "http://https://fake-network.com",
    },
  },
  "static": false,
}
`;

exports[`IcProxyPlugin should not add static assets if there is no output path 1`] = `
{
  "proxy": {
    "/api": {
      "changeOrigin": true,
      "pathRewrite": {
        "^/api": "/api",
      },
      "target": "http://https://fake-network.com",
    },
  },
  "static": [],
}
`;

exports[`IcProxyPlugin should not do anything if there is no network config 1`] = `undefined`;
